Root middleware handling (#36772)
vercel/next.js: errors/nested-middleware.md

> # Nested Middleware Deprecated

Reviewer: ```suggestion
# Nested Middleware
```

#### Why This Error Occurred

> You are defining a middleware file in a location different from `pages/_middleware` which is _deprecated_.

Reviewer: ```suggestion
You are defining a middleware file in a location different from `pages/_middleware` which is not allowed.
```

Declaring a middleware file under specific pages implied that it would _only_ be executed when pages below its declaration were matched.
This execution model allowed the nesting of multiple middleware, which is hard to reason about and led to consequences such as dragging effects between different middleware executions.

> The API has been _deprecated_ in favor of a simpler model with a single root middleware.

Reviewer: ```suggestion
The API has been removed in favor of a simpler model with a single root middleware.
```

#### Possible Ways to Fix It

To fix this error you must declare your middleware in the root pages folder and leverage `NextRequest` parsed URL to decide wether or not to execute the middleware code.

> For example, a middleware declared under `pages/about/_middleware.js` can be moved to `pages/_middleware` and updated so that it only executes when `about/*` matches:

Reviewer 2: ```suggestion
For example, a middleware declared under `pages/about/_middleware.js` can be moved to `pages/_middleware`. A conditional can be used to ensure the middleware executes only when it matches the `about/*` path:
```

```typescript
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  if (request.nextUrl.pathname.startsWith('/about')) {
    // Execute pages/about/_middleware.js
  }
}
```

> This also means that if you have more than one middleware you will need to reunite them in a single file and model their execution depending on the request.

Reviewer 2: ```suggestion
If you have more than one middleware, you will need to combine them into a single file and model their execution depending on the request.
```

> This also means that if you have more than one middleware you will need to reunite them in a single file and model their execution depending on the request.

Reviewer 2: I know it's probably not written yet. At the bottom, we can link to a changelog or post that has more context.
